File: python/design_pattern/template_method/main_template_method.py
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
Template Method Pattern

The Template Method pattern defines the skeleton of an algorithm in the
superclass but lets subclasses override specific steps of the algorithm
without changing its structure.

Key Components:
- AbstractClass: Defines template method and abstract operations
- ConcreteClass: Implements the abstract operations
"""

from abc import ABC, abstractmethod
from datetime import datetime


# Abstract Class with Template Method
class DataMiner(ABC):
    """
    Abstract class that defines the template method for data mining.
    The template method defines the algorithm's skeleton.
    """

    def mine(self, path: str) -> dict:
        """
        Template method - defines the algorithm skeleton.
        Subclasses cannot override this method.
        """
        # Fixed sequence of steps
        raw_data = self.extract_data(path)
        parsed_data = self.parse_data(raw_data)
        analyzed_data = self.analyze_data(parsed_data)
        report = self.generate_report(analyzed_data)

        # Hook methods - optional extensions
        self.send_notifications(report)

        return report

    @abstractmethod
    def extract_data(self, path: str) -> str:
        """Extract data from source - must be implemented by subclasses."""
        pass

    @abstractmethod
    def parse_data(self, raw_data: str) -> list[dict]:
        """Parse raw data into structured format - must be implemented."""
        pass

    def analyze_data(self, data: list[dict]) -> dict:
        """
        Analyze data - default implementation, can be overridden.
        """
        return {
            "record_count": len(data),
            "data": data,
        }

    def generate_report(self, analysis: dict) -> dict:
        """Generate report - default implementation, can be overridden."""
        return {
            "timestamp": datetime.now().isoformat(),
            "source_type": self.__class__.__name__,
            "analysis": analysis,
        }

    def send_notifications(self, report: dict) -> None:
        """Hook method - empty by default, can be overridden."""
        pass


# Concrete Classes
class CSVDataMiner(DataMiner):
    """Concrete class for mining CSV data."""

    def extract_data(self, path: str) -> str:
        print(f"  [CSVMiner] Extracting data from {path}")
        # Simulate reading CSV file
        return """name,age,city
Alice,30,New York
Bob,25,Los Angeles
Charlie,35,Chicago"""

    def parse_data(self, raw_data: str) -> list[dict]:
        print("  [CSVMiner] Parsing CSV data")
        lines = raw_data.strip().split("\n")
        headers = lines[0].split(",")
        result = []

        for line in lines[1:]:
            values = line.split(",")
            record = dict(zip(headers, values))
            result.append(record)

        return result

    def analyze_data(self, data: list[dict]) -> dict:
        print("  [CSVMiner] Analyzing CSV data")
        analysis = super().analyze_data(data)

        # Add CSV-specific analysis
        if data:
            ages = [int(d.get("age", 0)) for d in data if d.get("age", "").isdigit()]
            if ages:
                analysis["average_age"] = sum(ages) / len(ages)

        return analysis


class JSONDataMiner(DataMiner):
    """Concrete class for mining JSON data."""

    def extract_data(self, path: str) -> str:
        print(f"  [JSONMiner] Extracting data from {path}")
        # Simulate reading JSON file
        return '''[
            {"name": "Alice", "age": 30, "city": "New York"},
            {"name": "Bob", "age": 25, "city": "Los Angeles"},
            {"name": "Charlie", "age": 35, "city": "Chicago"}
        ]'''

    def parse_data(self, raw_data: str) -> list[dict]:
        print("  [JSONMiner] Parsing JSON data")
        import json

        return json.loads(raw_data)

    def send_notifications(self, report: dict) -> None:
        """Override hook to send notifications for JSON mining."""
        print("  [JSONMiner] Sending webhook notification")


class XMLDataMiner(DataMiner):
    """Concrete class for mining XML data."""

    def extract_data(self, path: str) -> str:
        print(f"  [XMLMiner] Extracting data from {path}")
        return """<people>
    <person><name>Alice</name><age>30</age><city>New York</city></person>
    <person><name>Bob</name><age>25</age><city>Los Angeles</city></person>
    <person><name>Charlie</name><age>35</age><city>Chicago</city></person>
</people>"""

    def parse_data(self, raw_data: str) -> list[dict]:
        print("  [XMLMiner] Parsing XML data")
        import xml.etree.ElementTree as ET

        root = ET.fromstring(raw_data)
        result = []

        for person in root.findall("person"):
            record = {}
            for child in person:
                record[child.tag] = child.text
            result.append(record)

        return result


class DatabaseMiner(DataMiner):
    """Concrete class for mining database data."""

    def extract_data(self, path: str) -> str:
        print(f"  [DBMiner] Connecting to database: {path}")
        # Simulate database query result
        return "Alice|30|New York\nBob|25|Los Angeles\nCharlie|35|Chicago"

    def parse_data(self, raw_data: str) -> list[dict]:
        print("  [DBMiner] Parsing query results")
        result = []
        for line in raw_data.strip().split("\n"):
            parts = line.split("|")
            result.append(
                {"name": parts[0], "age": parts[1], "city": parts[2]}
            )
        return result

    def analyze_data(self, data: list[dict]) -> dict:
        print("  [DBMiner] Running database-specific analysis")
        analysis = super().analyze_data(data)

        # Group by city
        cities: dict[str, int] = {}
        for record in data:
            city = record.get("city", "Unknown")
            cities[city] = cities.get(city, 0) + 1

        analysis["by_city"] = cities
        return analysis


# Report Generator example
class ReportGenerator(ABC):
    """Abstract class for generating reports."""

    def generate(self, title: str, data: list[dict]) -> str:
        """Template method for report generation."""
        report = []
        report.append(self.create_header(title))
        report.append(self.create_body(data))
        report.append(self.create_footer())
        return "\n".join(report)

    @abstractmethod
    def create_header(self, title: str) -> str:
        pass

    @abstractmethod
    def create_body(self, data: list[dict]) -> str:
        pass

    def create_footer(self) -> str:
        """Default footer - can be overridden."""
        return f"Generated on {datetime.now().strftime('%Y-%m-%d %H:%M')}"


class TextReport(ReportGenerator):
    """Generate plain text reports."""

    def create_header(self, title: str) -> str:
        return f"{'=' * 40}\n{title.upper()}\n{'=' * 40}"

    def create_body(self, data: list[dict]) -> str:
        if not data:
            return "No data available"

        lines = []
        for i, record in enumerate(data, 1):
            items = [f"{k}: {v}" for k, v in record.items()]
            lines.append(f"{i}. {', '.join(items)}")
        return "\n".join(lines)


class HTMLReport(ReportGenerator):
    """Generate HTML reports."""

    def create_header(self, title: str) -> str:
        return f"<html><head><title>{title}</title></head><body><h1>{title}</h1>"

    def create_body(self, data: list[dict]) -> str:
        if not data:
            return "<p>No data available</p>"

        html = ["<table border='1'>"]

        # Header row
        html.append("<tr>")
        for key in data[0].keys():
            html.append(f"<th>{key}</th>")
        html.append("</tr>")

        # Data rows
        for record in data:
            html.append("<tr>")
            for value in record.values():
                html.append(f"<td>{value}</td>")
            html.append("</tr>")

        html.append("</table>")
        return "".join(html)

    def create_footer(self) -> str:
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M")
        return f"<footer>Generated on {timestamp}</footer></body></html>"


class MarkdownReport(ReportGenerator):
    """Generate Markdown reports."""

    def create_header(self, title: str) -> str:
        return f"# {title}\n"

    def create_body(self, data: list[dict]) -> str:
        if not data:
            return "*No data available*"

        md = []

        # Header row
        headers = list(data[0].keys())
        md.append("| " + " | ".join(headers) + " |")
        md.append("| " + " | ".join(["---"] * len(headers)) + " |")

        # Data rows
        for record in data:
            values = [str(v) for v in record.values()]
            md.append("| " + " | ".join(values) + " |")

        return "\n".join(md)


def main() -> None:
    print("=" * 60)
    print("Template Method Pattern - Data Mining Demo")
    print("=" * 60)

    # Demo 1: Different data miners using same template
    print("\n--- CSV Data Mining ---")
    csv_miner = CSVDataMiner()
    result = csv_miner.mine("data/users.csv")
    print(f"Result: {result['analysis']['record_count']} records")
    if "average_age" in result["analysis"]:
        print(f"Average age: {result['analysis']['average_age']:.1f}")

    print("\n--- JSON Data Mining ---")
    json_miner = JSONDataMiner()
    result = json_miner.mine("data/users.json")
    print(f"Result: {result['analysis']['record_count']} records")

    print("\n--- XML Data Mining ---")
    xml_miner = XMLDataMiner()
    result = xml_miner.mine("data/users.xml")
    print(f"Result: {result['analysis']['record_count']} records")

    print("\n--- Database Mining ---")
    db_miner = DatabaseMiner()
    result = db_miner.mine("postgresql://localhost/users")
    print(f"Result: {result['analysis']['record_count']} records")
    print(f"By city: {result['analysis']['by_city']}")

    # Demo 2: Report generators
    print("\n" + "=" * 60)
    print("Template Method Pattern - Report Generation Demo")
    print("=" * 60)

    sample_data = [
        {"name": "Alice", "age": 30, "city": "New York"},
        {"name": "Bob", "age": 25, "city": "Los Angeles"},
        {"name": "Charlie", "age": 35, "city": "Chicago"},
    ]

    print("\n--- Text Report ---")
    text_gen = TextReport()
    print(text_gen.generate("User Report", sample_data))

    print("\n--- HTML Report ---")
    html_gen = HTMLReport()
    print(html_gen.generate("User Report", sample_data))

    print("\n--- Markdown Report ---")
    md_gen = MarkdownReport()
    print(md_gen.generate("User Report", sample_data))

    print("\n" + "=" * 60)
    print("Benefits of Template Method Pattern:")
    print("=" * 60)
    print("1. Defines algorithm skeleton in one place")
    print("2. Subclasses can override specific steps")
    print("3. Promotes code reuse")
    print("4. Controls extension points via hooks")
    print("5. Follows Hollywood Principle: 'Don't call us, we'll call you'")


if __name__ == "__main__":
    main()

File: rust/design_pattern/template_method/main_template_method.rs
// Template Method Design Pattern in Rust
// Demonstrates data mining with different file formats

/// DataMiner trait defines the template method pattern
/// The template method defines the algorithm skeleton
/// while allowing subclasses to override specific steps
trait DataMiner {
    // Template method - defines the algorithm skeleton
    // This method is final (not meant to be overridden)
    fn mine(&self) {
        println!("=== Starting Data Mining Process ===");
        self.open_file();
        let raw_data = self.extract_data();
        let processed_data = self.parse_data(&raw_data);
        self.analyze_data(&processed_data);
        self.send_report(&processed_data);
        self.hook_before_close();
        self.close_file();
        println!("=== Data Mining Complete ===\n");
    }

    // Abstract methods - must be implemented by concrete types
    fn open_file(&self);
    fn extract_data(&self) -> String;
    fn parse_data(&self, raw_data: &str) -> Vec<String>;
    fn close_file(&self);

    // Concrete methods with default implementation
    fn analyze_data(&self, data: &[String]) {
        println!("  [Analyze] Processing {} records", data.len());
        for (i, record) in data.iter().enumerate() {
            println!("    - Record {}: {}", i + 1, record);
        }
    }

    fn send_report(&self, data: &[String]) {
        println!("  [Report] Sending report with {} items", data.len());
    }

    // Hook method - empty by default, can be overridden
    fn hook_before_close(&self) {
        // Default: do nothing
    }
}

/// CSV Data Miner - concrete implementation for CSV files
struct CsvDataMiner {
    filename: String,
}

impl CsvDataMiner {
    fn new(filename: &str) -> Self {
        CsvDataMiner {
            filename: filename.to_string(),
        }
    }
}

impl DataMiner for CsvDataMiner {
    fn open_file(&self) {
        println!("  [Open] Opening CSV file: {}", self.filename);
    }

    fn extract_data(&self) -> String {
        println!("  [Extract] Reading CSV rows...");
        // Simulated CSV data
        String::from("name,age,city\nAlice,30,NYC\nBob,25,LA\nCharlie,35,Chicago")
    }

    fn parse_data(&self, raw_data: &str) -> Vec<String> {
        println!("  [Parse] Parsing CSV format...");
        let lines: Vec<&str> = raw_data.lines().collect();
        let header = lines[0];
        let fields: Vec<&str> = header.split(',').collect();

        lines[1..]
            .iter()
            .map(|line| {
                let values: Vec<&str> = line.split(',').collect();
                fields
                    .iter()
                    .zip(values.iter())
                    .map(|(k, v)| format!("{}={}", k, v))
                    .collect::<Vec<_>>()
                    .join(", ")
            })
            .collect()
    }

    fn close_file(&self) {
        println!("  [Close] Closing CSV file: {}", self.filename);
    }

    // Override hook to add CSV-specific cleanup
    fn hook_before_close(&self) {
        println!("  [Hook] CSV cleanup: clearing row buffer");
    }
}

/// JSON Data Miner - concrete implementation for JSON files
struct JsonDataMiner {
    filename: String,
    pretty_print: bool,
}

impl JsonDataMiner {
    fn new(filename: &str, pretty_print: bool) -> Self {
        JsonDataMiner {
            filename: filename.to_string(),
            pretty_print,
        }
    }
}

impl DataMiner for JsonDataMiner {
    fn open_file(&self) {
        println!("  [Open] Opening JSON file: {}", self.filename);
        if self.pretty_print {
            println!("    (Pretty print mode enabled)");
        }
    }

    fn extract_data(&self) -> String {
        println!("  [Extract] Reading JSON content...");
        // Simulated JSON data
        String::from(r#"[{"name":"Diana","score":95},{"name":"Eve","score":88}]"#)
    }

    fn parse_data(&self, raw_data: &str) -> Vec<String> {
        println!("  [Parse] Parsing JSON format...");
        // Simple manual JSON parsing for demonstration
        let mut results = Vec::new();
        let trimmed = raw_data.trim_start_matches('[').trim_end_matches(']');

        for obj in trimmed.split("},{") {
            let clean = obj
                .trim_start_matches('{')
                .trim_end_matches('}')
                .replace('"', "");

            let parts: Vec<&str> = clean.split(',').collect();
            let formatted: Vec<String> = parts
                .iter()
                .map(|p| p.replace(':', "="))
                .collect();

            results.push(formatted.join(", "));
        }
        results
    }

    fn close_file(&self) {
        println!("  [Close] Closing JSON file: {}", self.filename);
    }

    // Override the analyze method for JSON-specific analysis
    fn analyze_data(&self, data: &[String]) {
        println!("  [Analyze] JSON analysis of {} records:", data.len());
        for (i, record) in data.iter().enumerate() {
            if self.pretty_print {
                println!("    {{ Record {} }}", i + 1);
                for field in record.split(", ") {
                    println!("      - {}", field);
                }
            } else {
                println!("    Record {}: {}", i + 1, record);
            }
        }
    }
}

/// Game AI using Template Method - another example
trait GameAI {
    // Template method for AI turn
    fn take_turn(&self) {
        println!("--- AI Turn Started ---");
        self.collect_resources();
        self.build_structures();
        self.build_units();
        self.attack();
        println!("--- AI Turn Ended ---\n");
    }

    // Abstract methods
    fn build_structures(&self);
    fn build_units(&self);

    // Methods with default behavior
    fn collect_resources(&self) {
        println!("  [Resources] Collecting nearby resources");
    }

    fn attack(&self) {
        println!("  [Attack] Sending units to attack enemy");
    }
}

/// Aggressive Orc AI
struct OrcAI;

impl GameAI for OrcAI {
    fn build_structures(&self) {
        println!("  [Build] Constructing War Camp and Barracks");
    }

    fn build_units(&self) {
        println!("  [Units] Training Grunts and Wolfriders");
    }

    fn attack(&self) {
        println!("  [Attack] WAAAGH! All units charge!");
    }
}

/// Defensive Human AI
struct HumanAI;

impl GameAI for HumanAI {
    fn build_structures(&self) {
        println!("  [Build] Constructing Castle and Towers");
    }

    fn build_units(&self) {
        println!("  [Units] Training Knights and Archers");
    }

    fn collect_resources(&self) {
        println!("  [Resources] Peasants gathering gold and lumber");
    }
}

fn main() {
    println!("========================================");
    println!(" Template Method Pattern Demonstration");
    println!("========================================\n");

    // Data Mining Examples
    println!(">> DATA MINING EXAMPLES <<\n");

    let csv_miner = CsvDataMiner::new("users.csv");
    csv_miner.mine();

    let json_miner = JsonDataMiner::new("scores.json", true);
    json_miner.mine();

    // Game AI Examples
    println!(">> GAME AI EXAMPLES <<\n");

    let orc_ai = OrcAI;
    orc_ai.take_turn();

    let human_ai = HumanAI;
    human_ai.take_turn();

    println!("========================================");
    println!(" Pattern demonstration complete!");
    println!("========================================");
}
